sdks/go/pkg/beam/core/state/state.go

```diff
@@ -0,0 +1,117 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package state contains structs for reading and manipulating pipeline state.
+package state
+
+import (
+	"reflect"
+)
+
+type TransactionType_Enum int32
+
+const (
+	TransactionType_Set   TransactionType_Enum = 0
+	TransactionType_Clear TransactionType_Enum = 1
+)
+
+var (
+	ProviderType = reflect.TypeOf((*Provider)(nil)).Elem()
+)
+
+// TODO(#20510) - add other forms of state (MapState, BagState, CombiningState), prefetch, and clear.
+
+// Transaction is used to represent a pending state transaction. This should not be manipulated directly;
+// it is primarily used for implementations of the Provider interface to talk to the various State objects.
+type Transaction struct {
+	Key  string
+	Type TransactionType_Enum
+	Val  interface{}
+}
+
+// Provider represents the DoFn parameter used to get and manipulate pipeline state
+// stored as key value pairs (https://beam.apache.org/documentation/programming-guide/#state-and-timers).
+// This should not be manipulated directly. Instead it should be used as a parameter
+// to functions on State objects like state.Value.
+type Provider interface {
+	ReadValueState(id string) (interface{}, []Transaction, error)
+	WriteValueState(val Transaction) error
+}
+
+type PipelineState interface {
```

Comment:
Doc comment. Also fine if you are planning to add it in next PRs.

Reply:
Good call, I added one.

I may end up moving this to a different package when all is said and done (most likely when I have more execution focused state stuff) since I don't really want users implementing it. It will also eventually probably need some metadata about what state type should be used (ValueState in this case). Basically, its subject to change in my next PR(s) since it's more of an internals concept.